plugin/oauth-proxy.lua

```diff
@@ -49,27 +56,45 @@ local function from_hex(str)
     end))
 end
 
-local function decrypt_cookie(encrypted_cookie, encryption_key)
-    local encrypted = ngx.unescape_uri(encrypted_cookie)
+local function decrypt_cookie(encrypted_cookie, encryption_key_hex)
+
+    local all_bytes, err = base64.decode_base64url(encrypted_cookie)
+    if err then
+        ngx.log(ngx.WARN, 'A received cookie could not be base64url decoded ' .. err)
+        return nil
+    end
+
+    local min_size = VERSION_SIZE + GCM_IV_SIZE + 1 + GCM_TAG_SIZE
+    if #all_bytes < min_size then
+        ngx.log(ngx.WARN, 'A received cookie had an invalid length')
+        return nil
+    end
 
-    if not string.find(encrypted, ":") then
-        ngx.log(ngx.WARN, "Malformed cookie received with no valid separator")
+    local offset = 1
+    local version_byte = string.byte(all_bytes, offset, VERSION_SIZE)
+    if version_byte ~= CURRENT_VERSION then
+        ngx.log(ngx.WARN, 'A received cookie had invalid format')
         return nil
     end
 
-    local parts = split(encrypted, ":")
-    local iv = from_hex(parts[1])
-    local data = from_hex(parts[2])
+    offset = 1 + VERSION_SIZE
+    local iv_bytes = string.sub(all_bytes, offset, VERSION_SIZE + GCM_IV_SIZE)
+  
+    offset = 1 + VERSION_SIZE + GCM_IV_SIZE
+    local ciphertext_bytes = string.sub(all_bytes, offset, #all_bytes - GCM_TAG_SIZE)
 
-    local cipher = aes.cipher(256)
-    local aes_256_cbc_md5, err = aes:new(encryption_key, nil, cipher, { iv=iv })
+    offset = #all_bytes - GCM_TAG_SIZE + 1
+    local tag_bytes = string.sub(all_bytes, offset, #all_bytes)
```

Comment:
Shouldn't the last parameter be `GCM_TAG_SIZE`?

Reply:
In fact, I think that the third parameter can be ommitted and `sub` will take the rest of the string from `offset`.